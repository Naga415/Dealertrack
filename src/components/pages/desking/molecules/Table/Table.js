'use strict';
import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class Table extends React.PureComponent {
  constructor(props, context) {
    super(props, context);
  }

  applyFormatting = value => {
    const formatters = this.props.dataFormatters;
    if (formatters !== undefined) {
      const type = typeof value;
      const formatter = formatters[type];
      return formatter ? formatter(value) : value;
    }
    return value;
  }

  buildColumns = () => {
    return (
      <tr className="table-headers">
        {
          this.props.columns.map((column, i) =>
            <th key={i}>{column}</th>
          )
        }
      </tr>
    );
  }

  buildDataMap = (row) => {
    const keys = Object.keys(this.props.dataMap);
    const rowData = [];
    keys.forEach((key, i) => {
      rowData[i] = row[key];
    });
    return rowData;
  }

  buildRows = (row) => {
    const mappedRow = this.buildDataMap(row);
    return (
      mappedRow.map((value, i) =>
        <td key={i}>{this.applyFormatting(value)}</td>
      )
    );
  }

  render() {
    const { htmlId, title, data } = this.props;
    const columns = this.buildColumns();
    const rows = data.map((row, i) =>
      <tr key={i}>{this.buildRows(row)}</tr>
    );
    return (
      <div id={htmlId} className="desking-table">
        <span className="desking-table__title">
          {title}
        </span>
        <table>
          <thead>
            {columns}
          </thead>
          <tbody>
            {rows}
          </tbody>
        </table>
      </div>
    );
  }
}

Table.propTypes = {
  htmlId: PropTypes.string,
  columns: PropTypes.arrayOf(PropTypes.string),
  data: PropTypes.arrayOf(PropTypes.object),
  /** An array that maps the columns with the data
   * [columnName: 'DisplayName'}]
   */
  dataMap: PropTypes.object,
  title: PropTypes.string,
  /** Holds the different data formatters that needs to be applied on the [data] array
   * formatter = (value) => {}
   * {{Type: formatter}} e.g. {'number': formatter}
   */
  dataFormatters: PropTypes.object
};

export default Table;
