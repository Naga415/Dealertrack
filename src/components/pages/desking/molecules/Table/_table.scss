$deskingTable:'.desking-table'; // namespace for component

#{$deskingTable} {
  width: 100%;
  padding: 10px;

  &__title {
    font-size: 16px;
    font-weight: 400;
    display: block;
    padding-bottom: 15px;
  }

  table {
    width: 100%;

    thead {
      tr {
        th {
          font-size: 14px;
          font-weight: 500;
        }
      }
    }

    tbody {
      tr {
        td {
          font-size: 13px;
        }
      }
    }
  }
}
