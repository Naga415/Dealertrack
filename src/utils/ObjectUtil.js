'use strict';
export default class ObjectUtil {

  static clone(objectToClone, propsToMerge) {
    return Object.assign(Object.create(Object.getPrototypeOf(objectToClone)), objectToClone, propsToMerge);
  }

  static cloneArray(array) {
    if (!array) {
      return array;
    }
    return array.map(o => this.clone(o));
  }
}
