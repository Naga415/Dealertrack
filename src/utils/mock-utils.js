export function generateNumArray(count){
  const numberArray = [];
  for (let i = 1; i <= count; i++) { numberArray.push(i); }
  return numberArray; 
}