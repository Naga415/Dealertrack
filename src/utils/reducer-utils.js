import extend from 'lodash/extend';
import clone from 'lodash/clone';

export function reduceAtSubState(rootState, action, subStatePath, subStateReducer){
  return subStatePath.reverse().reduce((p, n, i) => {
    if (i === 0){
      return extend({}, p, {
        [n]: subStateReducer(getSubState(rootState, subStatePath), action)
      });
    }
    return extend({},
      getSubState(rootState, clone(subStatePath).slice(0, subStatePath.length - i)),
      {
        [n]: p
      });
  }, getSubState(rootState, clone(subStatePath).slice(0, subStatePath.length - 1)));
}

export function getSubState(rootState, subStatePath){
  if (!subStatePath.length){
    return rootState;
  }

  return subStatePath.reduce((p, n) => p[n], rootState);
}

export function convertNullOrUndefinedToEmptyStrings(source, targetKeys){
  const result = clone(source);
  targetKeys.forEach(k => !result[k] && ( result[k] = ""));
  return result;
}
