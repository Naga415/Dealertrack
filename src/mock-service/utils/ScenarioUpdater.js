'use strict';
import * as ScenarioTypes from '../../models/ScenarioTypes';
import ObjectUtil from '../../utils/ObjectUtil';
import '../../models/FinanceSummary';

export default class ScenarioUpdater {
  static update(scenario){
    if (!scenario){
      return scenario;
    }

    switch (scenario.type){
      case ScenarioTypes.FINANCE:
        return this.updateFinance(scenario);
      case ScenarioTypes.CASH:
        return this.updateCash(scenario);
      case ScenarioTypes.LEASE:
        return this.updateLease(scenario);
      case ScenarioTypes.BALLOON:
        return this.updateBalloon(scenario);
    }
  }

  static updateFinance(scenario){
    const msrp=scenario.details.msrp;
    const sellingPrice =this.numOrDefault(scenario.details.sellingPrice,0);
    const addOns =this.numOrDefault(scenario.details.addOns,0);
    const fees=this.numOrDefault(scenario.details.fees,0);
    const taxes=this.numOrDefault(scenario.details.taxes,0);
    const lenderFee=this.numOrDefault(scenario.details.lenderFee,0);
    const rebate=this.numOrDefault(scenario.details.rebate,0);
    const rate=this.numOrDefault(scenario.details.rate,0);
    const term=this.numOrDefault(scenario.details.term,0);
    const trade = scenario.trade;
    const downpayment = this.numOrDefault(scenario.details.downpayments[0], 0);
    const amountFinanced = sellingPrice - rebate - downpayment;

    const updatedScenario = ObjectUtil.clone(scenario);
    updatedScenario.summary =ObjectUtil.clone(scenario.summary);
    updatedScenario.summary.amountFinanced = amountFinanced;
    updatedScenario.summary.amountDueAtSigning = amountFinanced * .5;
    updatedScenario.summary.balanceDue = taxes + fees + lenderFee - rebate - downpayment;
    updatedScenario.summary.holdDeposit = null;
    updatedScenario.summary.purchasePriceWithAdds = sellingPrice + addOns;
    updatedScenario.summary.discountFromSellingPrice = sellingPrice-msrp;
    updatedScenario.summary.netTrade = this.numOrDefault(trade.allowance,0) - this.numOrDefault(trade.payoff,0);
    updatedScenario.summary.interestRateTotal = null;
    updatedScenario.summary.totalPayments = term * 12;
    updatedScenario.summary.subtotal = sellingPrice + fees + addOns - rebate - downpayment;
    updatedScenario.summary.payment = this.getPayment(amountFinanced, rate, term);

    return updatedScenario;
  }

  static getPayment(principle, rate,term){
    try {
      const effectiveRate = ((rate/100)/12);
      const numberOfPayments = term * 12;
      return principle*(effectiveRate/(1-Math.pow((1+effectiveRate),(-1*numberOfPayments))));
    }
    catch (e){
      return 0;
    }
  }

  static updateCash(scenario){
    return scenario;
  }

  static updateLease(scenario){
    return scenario;
  }

  static updateBalloon(scenario){
    return scenario;
  }
  static numOrDefault(number, defaultNumber){
    return (number===0||(number&&!isNaN(number)))? parseFloat(number,10): defaultNumber;
  }
}
