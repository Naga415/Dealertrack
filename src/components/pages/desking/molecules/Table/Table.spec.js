import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Table from './Table';
import { getFormattedNumber } from '../../../../../utils/numberFormatter';

describe('<Table/> component', () => {
  const columns = ["Column1", "Column2", "Column3"];
  const data = [
    {
      col1: 'Col1 value',
      col2: 1235,
      col3: 2450
    },
    {
      col1: 7853,
      col2: 'col2 value',
      col3: 78552
    }
  ];
  const dataMap = {
    col1: "Column1",
    col2: "Column2",
    col3: "Column3",
  };

  it('renders <Table/> that matches the data', () => {
    const component = renderer.create(<Table htmlId="NewTable" title="Back End Gross" data={data} columns={columns} dataMap={dataMap} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders <Table> with a formatted data when custom formatter are passed into it', () => {
    const dataFormatters = { 'number': getFormattedNumber };
    const component = shallow(<Table htmlId="NewTable" title="Back End Gross" data={data} columns={columns} dataMap={dataMap} dataFormatters={dataFormatters} />);
    expect(component.find('tbody tr').length).toBe(2);
    expect(component.find('tbody tr td').at(0).text()).toEqual('Col1 value');
    expect(component.find('tbody tr td').at(1).text()).toEqual('1,235');
    expect(component.find('tbody tr td').at(2).text()).toEqual('2,450');
  });
});
