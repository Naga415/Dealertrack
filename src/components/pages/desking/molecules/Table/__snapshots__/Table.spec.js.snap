// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table/> component renders <Table/> that matches the data 1`] = `
<div
  className="desking-table"
  id="NewTable"
>
  <span
    className="desking-table__title"
  >
    Back End Gross
  </span>
  <table>
    <thead>
      <tr
        className="table-headers"
      >
        <th>
          Column1
        </th>
        <th>
          Column2
        </th>
        <th>
          Column3
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Col1 value
        </td>
        <td>
          1235
        </td>
        <td>
          2450
        </td>
      </tr>
      <tr>
        <td>
          7853
        </td>
        <td>
          col2 value
        </td>
        <td>
          78552
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
